src/shogun/evaluation/CrossValidatedCalibration.cpp

```diff
@@ -0,0 +1,406 @@
+
+	CBinaryLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values;
+	CBinaryLabels* result_labels;
+
+	for (index_t i = 0; i < num_machines; ++i)
+	{
+		temp_machine = (CMachine*)m_calibration_machines->get_element(i);
+		temp_result = (CBinaryLabels*)temp_machine->apply(features);
+		if (i == 0)
+		{
+			result_values = temp_result->get_values();
+			result_labels = temp_result;
+		}
+		else
+		{
+			result_values += temp_result->get_values();
+		}
+	}
+
+#pragma omp parallel for
+	for (index_t i = 0; i < result_values.vlen; ++i)
+	{
+		result_values[i] = result_values[i] / num_machines;
+	}
+
+	result_labels->set_values(result_values);
+
+	return result_labels;
+}
+
+CBinaryLabels* CCrossValidatedCalibration::apply_locked_binary(
+    SGVector<index_t> subset_indices)
+{
+
+	index_t num_machines = m_calibration_machines->get_num_elements();
+
+	CBinaryLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values = NULL;
+	CBinaryLabels* result_labels;
+
+	for (index_t i = 0; i < num_machines; ++i)
+	{
+		temp_machine = (CMachine*)m_calibration_machines->get_element(i);
+		temp_result =
+		    (CBinaryLabels*)temp_machine->apply_locked(subset_indices);
+		if (result_values == NULL)
+		{
+			result_values = temp_result->get_values();
+			result_labels = temp_result;
+		}
+		else
+		{
+			result_values += temp_result->get_values();
+		}
+	}
+
+#pragma omp parallel for
+	for (index_t i = 0; i < result_values.vlen; ++i)
+	{
+		result_values[i] = result_values[i] / num_machines;
+	}
+
+	result_labels->set_values(result_values);
+
+	return result_labels;
+}
+
+CMulticlassLabels*
+CCrossValidatedCalibration::apply_multiclass(CFeatures* features)
+{
+	index_t num_machines = m_calibration_machines->get_num_elements();
+
+	index_t num_classes = ((CMulticlassLabels*)m_labels)->get_num_classes();
+
+	CMulticlassLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values, temp_values;
+	CMulticlassLabels* result_labels;
+
+	for (index_t i = 0; i < num_machines; ++i)
+	{
+		temp_machine = (CMachine*)m_calibration_machines->get_element(i);
+		temp_result = (CMulticlassLabels*)temp_machine->apply(features);
+		if (i == 0)
+		{
+			result_labels = temp_result;
+		}
+		else
+		{
+			for (index_t j = 0; j < num_classes; ++j)
+			{
+				result_values = temp_result->get_multiclass_confidences(j);
+				temp_values = result_labels->get_multiclass_confidences(j);
+				temp_values += result_values;
+				result_labels->set_multiclass_confidences(j, temp_values);
+			}
+		}
+	}
+
+	for (index_t i = 0; i < num_classes; ++i)
+	{
+		temp_values = result_labels->get_multiclass_confidences(i);
+
+#pragma omp parallel for
+		for (index_t j = 0; j < temp_values.vlen; j++)
+		{
+			temp_values[j] = temp_values[j] / num_machines;
+		}
+		result_labels->set_multiclass_confidences(i, temp_values);
+	}
+
+	return result_labels;
+}
+
+CMulticlassLabels* CCrossValidatedCalibration::apply_locked_multiclass(
```

Comment:
almost the same as `CCrossValidatedCalibration::apply_multiclass` so plz define a template or a function where you pas the apply method